README.md
# Linux.I2C.MCP342X, driver for A/D

[![Clojars
Project](https://img.shields.io/clojars/v/io.helins/linux.i2c.mcp342x.svg)](https://clojars.org/io.helins/linux.i2c.mcp342x)

[![Cljdoc](https://cljdoc.org/badge/io.helins/linux.i2c.mcp342x)](https://cljdoc.org/d/io.helins/linux.i2c.mcp342x)

Clojure library for talking to the MCP342x family of A/D converters using
[I2C](https://en.wikipedia.org/wiki/I%C2%B2C) :

- MCP3421
- MCP3422
- MCP3423
- MCP3424
- MCP3425
- MCP3426
- MCP3427
- MCP3428

Relies on [helins/linux.i2c](https://github.com/helins/linux.i2c.clj).


## Usage

This is an overview.

The [detailed API is available on Cljdoc](https://cljdoc.org/d/io.helins/linux.i2c.mcp342x).

In short, without error handling :

```clj
(require '[helins.linux.i2c         :as i2c]
         '[helins.linux.i2c.mcp342x :as mcp342x])


(with-open [bus (i2c/bus "/dev/i2c-1")]

  (i2c/select-slave bus
                    (mcp342x/address true
                                     false
                                     true))

  (mcp342x/configure bus
                     {:mcp342x/channel    2
                      :mcp342x/mode       :continuous
                      :mcp342x/pga        :x1
                      :mcp342x/resolution :16-bit})

  (mcp342x/read-channel bus
                        :16-bit)

  => {:mcp342x/channel    2
      :mcp342x/micro-volt 913000
      :mcp342x/mode       :continuous
      :mcp342x/pga        :x1
      :mcp342x/resolution :16-bit})
```


## License

Copyright © 2017 Adam Helinski

Licensed under the term of the Mozilla Public License 2.0, see LICENSE.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.



## [Unreleased]

### Added

### Changed

### Fixed



## [2.0.0-beta0] - 2021-03-23

### Changed

- Namespace parameters under "mcp342x"
- Relicensed under MPL 2.0
- Renamed group id



## [1.0.0-beta1] - 2018-06-14

### Changed

- Improved docstrings



## [1.0.0-beta0] - 2018-06-13

### Changed

- Upgraded underlying I2C library



## [0.0.1] - 2018-03-4

### Changed

- Improved docstrings



## [0.0.0] - 2018-03-03

### Added

- First API iteration



[Unreleased]: https://github.com/helins/linux.i2c.mcp342x.clj/compare/2.0.0-beta0...HEAD
[2.0.0-beta0]: https://github.com/helins/linux.i2c.mcp342x.clj/compare/1.0.0-beta1...2.0.0-beta0
[1.0.0-beta1]: https://github.com/helins/linux.i2c.mcp342x.clj/compare/1.0.0-beta0...1.0.0-beta1
[1.0.0-beta0]: https://github.com/helins/linux.i2c.mcp342x.clj/compare/0.0.1...1.0.0-beta0
[0.0.1]: https://github.com/helins/linux.i2c.mcp342x.clj/compare/0.0.0...0.0.1
[0.0.0]: https://github.com/helins/linux.i2c.mcp342x.clj/tree/0.0.0


project.clj
(defproject dvlopt/linux.i2c.mcp342x
            "1.0.0-beta1"

  :description  "Talk to the MCP342x family of ADC via I2C"
  :url          "https://github.com/dvlopt/linux.i2c.mcp342x"
  :license      {:name "Eclipse Public License"
                 :url  "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[dvlopt/linux.i2c "1.1.0"]
                 [dvlopt/void      "0.0.0"]]
  :profiles     {:dev {:source-paths ["dev"]
                       :main         user
                       :dependencies [[criterium              "0.4.4"]
                                      [org.clojure/clojure    "1.9.0"]
                                      [org.clojure/test.check "0.10.0-alpha2"]]
                       :plugins      [[lein-codox      "0.10.3"]
                                      [venantius/ultra "0.5.2"]]
                       :codox        {:output-path  "doc/auto"
                                      :source-paths ["src"]}
                       :repl-options {:timeout 180000}
                       :global-vars  {*warn-on-reflection* true}}})


src/mcp342x/core.clj
(ns mcp342x.core

  "Utilities for understanding the I2C protocol for the MCP342x family of A/D."

  {:author "Adam Helinski"})




;;;;;;;;;; Misc


(defn address

  "Gets the address of a slave.
  
   The user can use pins `a0`, `a1` and `a2` to modify the default address."

  ([]

   0x0d)


  ([a0 a1 a2]

   (bit-or (address)
           (if a0
             0x01
             0x00)
           (if a1
             0x02
             0x00)
           (if a2
             0x04
             0x00))))




;;;;;;;;;; Configuration byte


(def config-flags

  "Bit flags for the configuration byte.
  
   Cf. `config-flag`"

  {:ready?  {true  0x80
             false 0x00}
   :channel {1 0x00
             2 0x20
             3 0x40
             4 0x60}
   :mode    {:continuous 0x10
             :one-shot   0x00}
   :bits    {12 0x00
             14 0x04
             16 0x08
             18 0x0c}
   :pga     {1 0x00
             2 0x01
             4 0x02
             8 0x03}})




(defn config-flag

  "Given a parameter and a value, gets the related bit flag.
  
   Throws an IllegalArgumentException when not found.
  
  
   Cf. `config-flags` for parameters and values."

  [parameter value]

  (or (get-in config-flags
              [parameter value])
      (throw (IllegalArgumentException. (format "Bit flag not found for parameter '%s' with value '%s'."
                                                parameter
                                                value)))))




(defn from-config

  "Given a configuration byte, returns an understandable map.

     {:ready?
       When true,  the newest value has not been read because it is not ready yet.
            false, the newest value has been read.
       Might be a bit counter-intuitive, but we keep the original semantics.

      :channel
       Selected channel.
  
      :mode
       Either :continuous, meaning it is constantly measuring.
              :one-shot,   meaning it only measure when the master writes a configuration byte
                           with `:ready` set to true or any other bit is different from the last
                           time.

      :bits
       Selected precision, one of #{12 14 16 18} bits.

      :pga
       Selected Programmable Gain Amplifier, one of #{1 2 4 8}.}
  

  Cf. `config-flags`"

  [b]

  {:ready?  (bit-test b
                      7)
   :channel (if (bit-test b
                          5)
              (if (bit-test b
                            6)
                4
                2)
              (if (bit-test b
                            6)
                3
                1))
   :mode    (if (bit-test b
                          4)
              :continuous
              :one-shot)
   :bits    (if (bit-test b
                          2)
              (if (bit-test b
                            3)
                18
                14)
              (if (bit-test b
                            3)
                16
                12))
   :pga     (if (bit-test b
                          0)
              (if (bit-test b
                            1)
                8
                2)
              (if (bit-test b
                            1)
                4
                1))})




(defn to-config

  "Given a configuration map, returns a configuration byte.
  
     {:ready? 
       In :continuous mode, doesn't do anything.
          :one-shot   mode, when 'true', signals the slave it should measure once.

      :channel
       Which channel select.

      :mode
       Either :continuous for measuring constantly.
              :one-shot   for measuring only once when asked to.

      :bits
       Precision, one of #{12 14 16 18} bits.

      :pga
       Programmable Gain Amplifier, one of #{1 2 4 8}.}
  
  
   Cf. `config-flags`"
  
  ([]

   0x90)


  ([{:as   config
     :keys [ready?
            channel
            mode
            bits
            pga]
     :or   {ready?  true 
            channel 1
            mode    :continuous
            bits    12
            pga     1}}]
 
   (bit-or (config-flag :ready?
                        ready?)
           (config-flag :channel
                        channel)
           (config-flag :mode
                        mode)
           (config-flag :bits
                        bits)
           (config-flag :pga
                        pga))))




;;;;;;;;;; Output code


(defn- -ubyte

  "Given a byte array, gets an 'unsigned' byte and shifts it to the left if needed."

  ([ba i]

   (-ubyte ba
           i
           0xff))


  ([^bytes ba i mask]

   (bit-and mask
            (aget ba
                  i)))


  ([^bytes ba i mask shift-left]

   (bit-shift-left (-ubyte ba
                           i
                           mask)
                   shift-left)))




(defn- -sign

  "Given a byte, checks the bit at `i` and treats it as a sign bit.
  
   Returns a signed byte accordingly."

  [i b]

  (if (bit-test b
                i)
    (- b)
    b))




(defn- -output-code-2bytes

  "Helper for combining 2 bytes into an ouput code which can then be used to
   compute the input voltage."

  [mask msb ^bytes ba]

  (-sign msb
         (bit-or (-ubyte ba
                         1)
                 (-ubyte ba
                         0
                         mask
                         8))))




(defn output-code-12bits

  "Given 2 bytes, computes the output code in 12 bits mode.
  
   Takes care of the sign."

  [^bytes ba]

  (-output-code-2bytes 0x0f
                       11
                       ba))




(defn output-code-14bits 

  "Given 2 bytes, computes the output code in 14 bits mode.
  
   Takes care of the sign."

  [^bytes ba]

  (-output-code-2bytes 0x3f
                       13
                       ba))




(defn output-code-16bits

  "Given 2 bytes, computes the output code in 16 bits mode.
  
   Takes care of the sign."

  [^bytes ba]

  (-output-code-2bytes 0xff
                       15
                       ba))




(defn output-code-18bits

  "Given 3 bytes, computes the output code in 18 bits mode.
  
   Takes care of the sign."

  [^bytes ba]

  (-sign 17
         (bit-or (-ubyte ba
                         2)
                 (-ubyte ba
                         1
                         0xff
                         8)
                 (-ubyte ba
                         0
                         0x03
                         16))))




(defn output-code

  "Given a bit mode (12, 14, 16 or 18) and a byte array (size 3 for 18 bits, size 2 otherwise),
   computes the output code which can then be converted to the input voltage."

  [^long bits ba]

  ((case bits
     12 output-code-12bits
     14 output-code-14bits
     16 output-code-16bits
     18 output-code-18bits) ba))




;;;;;;;;;; Read data


(defn input-voltage

  "Given a bit mode (12, 14, 16 or 18), the PGA (1, 2, 4 or 8) and an output code,
   computes the input voltage."

  [^long bits ^long pga ^long output-code]

  (let [lsb (case bits
              12 1000
              14  250
              16   62.5
              18   15.625)]
    (/ (* output-code
          lsb)
       pga)))




(defn from-read

  "Given bytes read from the slave, returns a map containing the configuration and the input voltage
   expressed in µV.
  

   In 12, 14 or 16 bits mode -> 3 bytes (2 data bytes + the configuration byte).
  
   In 18 bits mode           -> 4 bytes (3 data bytes + the configuration byte)."

  [^bytes ba]

  (let [{:as   config
         :keys [bits
                pga]} (from-config (last ba))]
    (assoc config
           :micro-volt
           (input-voltage bits
                          pga
                          (output-code bits
                                       ba)))))



;;;;;;;;;;


;; TODO general call ? Cf. Any datasheet 5.4


src/main/helins/linux/i2c/mcp342x.clj
;; This Source Code Form is subject to the terms of the Mozilla Public
;; License, v. 2.0. If a copy of the MPL was not distributed with this
;; file, You can obtain one at https://mozilla.org/MPL/2.0/.



(ns helins.linux.i2c.mcp342x

  "Library for talking to MCP342x converters using I2C.
  
   Those devices are configured using a single byte and the input voltage is read directly.
  
   Parameters used throughout this library are :

     :mcp342x/channel
       Up to 4 channels, from 1 to 4, are available depending on the model.

     :mcp342x/converting?
       When writing the configuration byte, this parameter must be set to true for initiating
       a new measure in :one-shot mode. It does not matter in :continuous mode.
       When reading the input voltage, the converter sets this parameter to false when a new
       conversion is ready.

     :mcp342x/mode
       In :continuous mode, the converter measures the input voltage constantly whereas in
       :one-shot mode, the measure only happens when the master writes a configuration byte
       with :mcp342x/converting? set to true.

     :mcp342x/pga
       Programamble Gain Amplifier.
       #{:x1 :x2 :x4 :x8}

     :mcp342x/resolution
       Number of bits the input voltage is represented by, depending on the model.
       #{:12-bit :14-bit :16-bit :18-bit}
  

   IO operations will throw in case of failure and are performed using this library :
  
     https://github.com/helins/linux.i2c.clj"

  {:author "Adam Helinski"}

  (:require [helins.linux.i2c :as i2c]))


;;;;;;;;;; Misc


(def default+

  "Default values for this namespace."

  {:mcp342x/channel     1
   :mcp342x/converting? true
   :mcp342x/mode        :continuous
   :mcp342x/pga         :x1
   :mcp342x/resolution  :12-bit})


;;;;;;;;;; Bit manipulations


(defn- -sign

  ;; Given a byte, checks the bit at `i` and treats it as a sign bit.
  ;;
  ;; Returns a signed byte accordingly.

  [i b]

  (if (bit-test b
                i)
    (- b)
    b))



(defn- -ubyte

  ;; Given a sequence of bytes, gets an 'unsigned' byte and shifts it to the left if needed.

  ([bs i mask]

   (bit-and (nth bs
                 i)
            mask))

  ([bs i mask shift-left]

   (bit-shift-left (-ubyte bs
                           i
                           mask)
                   shift-left)))


;;;;;;;;;; Handling parameters


(def ^:private -flags

  ;; Bit flags for the configuration byte.

  {:mcp342x/channel     {1 0x00
                         2 0x20
                         3 0x40
                         4 0x60}
   :mcp342x/converting? {true  0x80
                         false 0x00}
   :mcp342x/mode        {:continuous 0x10
                        :one-shot   0x00}
   :mcp342x/pga         {:x1 0x00
                         :x2 0x01
                         :x4 0x02
                         :x8 0x03}
   :mcp342x/resolution  {:12-bits 0x00
                         :14-bits 0x04
                         :16-bits 0x08
                         :18-bits 0x0c}})




(defn- -param+->byte

  ;; Given a map of parameters, returns a configuration byte.
  
  [param+]

  (reduce (fn add-flag [b param]
            (let [value (or (get param+
                                 param)
                            (get default+
                                 param))]
              (bit-or b
                      (or (get-in -flags
                                  [param value])
                          (throw (IllegalArgumentException. (format "Required configuration is invalid : %s %s"
                                                                    param
                                                                    value)))))))
          0
          [:mcp342x/channel
           :mcp342x/converting?
           :mcp342x/mode
           :mcp342x/pga
           :mcp342x/resultion]))



(defn- -byte->param+

  ;; Converts a configuration byte into a map of parameters.

  [b]

  {:mcp342x/channel     (if (bit-test b
                                      5)
                          (if (bit-test b
                                        6)
                            4
                            2)
                          (if (bit-test b
                                        6)
                            3
                            1))
   :mcp342x/converting? (bit-test b
                                  7)
   :mcp342x/mode        (if (bit-test b
                                      4)
                         :continuous
                         :one-shot)
   :mcp342x/pga         (if (bit-test b
                                      0)
                          (if (bit-test b
                                        1)
                            :x8
                            :x2)
                          (if (bit-test b
                                        1)
                            :x4
                            :x1))
   :mcp342x/resolution  (if (bit-test b
                                      2)
                          (if (bit-test b
                                        3)
                            :18-bit
                            :14-bit)
                          (if (bit-test b
                                        3)
                            :16-bit
                            :12-bit))})


;;;;;;;;;; Computing output codes and input voltages


(defn- -compute-output-code

  ;; Helper for combining 2 bytes into an ouput code which can then be used to
  ;; compute the input voltage.

  [bs msb mask]

  (-sign msb
         (bit-or (get bs
                      1)
                 (-ubyte bs
                         0
                         mask
                         8))))



(defn- -output-code

  ;; Computes the output-code taking into account the resolution.

  [param+ bs]

  (condp identical?
         (:mcp342x/resolution param+)
    :12-bit (-compute-output-code bs
                                  11
                                  0x0f)
    :14-bit (-compute-output-code bs
                                  13
                                  0x3f)
    :16-bit (-compute-output-code bs
                                  15
                                  0xff)
    :18-bit (-sign 17
                   (bit-or (nth bs
                                2)
                           (-ubyte bs
                                   1
                                   0xff
                                   8)
                           (-ubyte bs
                                   0
                                   0x03
                                   16)))))



(defn- -input-voltage

  ;; Given the resolution, the PGA and the output code, computes the input voltage."

  ^double

  [param+ bs]

  (let [^double lsb         (condp identical?
                                   (:mcp342x/resolution param+)
                              :12-bits 1000
                              :14-bits  250
                              :16-bits   62.5
                              :18-bits   15.625)
                output-code (-output-code param+
                                          bs)
        ^long   pga         (condp identical?
                                   (:mcp342x/pga param+)
                              :x1 1
                              :x2 2
                              :x4 4
                              :x8 8)]
    (double (/ (* output-code
                  lsb)
               pga))))


;;;;;;;;;; API


(defn address

  "Returns the address of a slave device.
  
   Pins `a0`, `a1` and `a2` are used to modify the default address by specifying a boolean value for each."

  ([]

   0x0d)


  ([a0 a1 a2]

   (bit-or (address)
           (if a0
             0x01
             0x00)
           (if a1
             0x02
             0x00)
           (if a2
             0x04
             0x00))))



(defn configure

  "Configures the slave device by providing parameters.

   Do not forget to select the slave device first.
  
   In absence of a parameter, a default value will be used.
  
   See [[default+]]."

  [bus param+]

  (i2c/write bus
             [(-param+->byte param+)])
  nil)



(defn read-channel

  "Reads a channel and returns a map containing the parameters under which the measure has been done as well as the
   current input voltage under :mcp342x/micro-volt."

  ([bus]

   (read-channel bus
                 (get default+
                      :mcp342x/resolution)))


  ([bus resolution]

   (let [result (i2c/read bus
                          (condp identical?
                                 resolution
                                 :12-bit 3
                                 :14-bit 3
                                 :16-bit 3
                                 :18-bit 4))
         param+ (-byte->param+ (last result))]
     (assoc param+
            :mcp342x/micro-volt
            (-input-voltage param+
                            result)))))


;;;;;;;;;;


;; TODO general call ? Cf. Any datasheet 5.4
